### PR TITLE
#789 bounties

### PR DESCRIPTION
**Pull request by almondy22**

### CODE REVIEW

File: src/contracts/peeranha.main/question_container.hpp [hunk 1]
```diff
@@ -28,6 +28,8 @@
 #define PROPERTY_MODERATION_VOTES 2
 #define PROPERTY_LAST_MODIFIED 3
 #define PROPERTY_QUESTION_TYPE 4
+#define PROPERTY_BOUNTY_COUNT 5
```
Comment (sergeyi23) [inline src/contracts/peeranha.main/question_container.hpp:31]: Let's use values 14 and 15. I am not sure why these 2 values were skipped. It is possible that they were used for something else.

File: src/contracts/peeranha.token/token.cpp [hunk 2]
```diff
@@ -256,6 +256,38 @@ void token::inviteuser(name inviter, name invited_user) {
...
+        a.question_id = question_id;
+        a.timestamp = timestamp;
+        a.status = BOUNTY_STATUS_ACTIVE;
+    });
+}
+
+// remove scope of user and make it with global scope, f.e. "all_bounty_scope"
+// remove questionUser because it's not needed and needs authorization
+
+void token::givebounty(name user, uint64_t question_id) {
```
Comment (sergeyi23) [inline src/contracts/peeranha.token/token.cpp:278]: Remove commented code

File: src/contracts/peeranha.token/token.cpp [hunk 3]
```diff
@@ -256,6 +256,38 @@ void token::inviteuser(name inviter, name invited_user) {
...
+    bountytable.emplace(user, [&](auto &a) {
+        a.user = user;
+        a.bounty = bounty;
+        a.question_id = question_id;
+        a.timestamp = timestamp;
+        a.status = BOUNTY_STATUS_ACTIVE;
+    });
+}
+
+// remove scope of user and make it with global scope, f.e. "all_bounty_scope"
```
Comment (sergeyi23) [inline src/contracts/peeranha.token/token.cpp:275]: Are these comments needed?

File: src/contracts/peeranha.token/token.cpp [hunk 4]
```diff
@@ -256,6 +256,38 @@ void token::inviteuser(name inviter, name invited_user) {
       });
 }
 
+void token::getbounty(name user, asset bounty, uint64_t question_id, uint64_t timestamp) {
```
Comment (sergeyi23) [inline src/contracts/peeranha.token/token.cpp:259]: The name not very clear.
Maybe `setbounty`?

File: src/contracts/peeranha.token/token.cpp (hunk 2 above)
Comment (sergeyi23) [inline src/contracts/peeranha.token/token.cpp:278]: Name not very clear. Maybe `paybounty`?

File: src/contracts/peeranha.token/token.hpp [hunk 5]
```diff
@@ -49,8 +49,17 @@ class[[eosio::contract("peeranha.token")]] token : public contract {
 
   [[eosio::action]] void pickupreward(name user, const uint16_t period);
 
+  // Get bounty from user who asked question
```
Comment (sergeyi23) [inline src/contracts/peeranha.token/token.hpp:52]: Remove commented code

File: src/contracts/peeranha.token/token.cpp [hunk 6]
```diff
@@ -256,6 +256,38 @@ void token::inviteuser(name inviter, name invited_user) {
...
+
+// remove scope of user and make it with global scope, f.e. "all_bounty_scope"
+// remove questionUser because it's not needed and needs authorization
+
+void token::givebounty(name user, uint64_t question_id) {
+    require_auth(user);
+//    name peertken = name("peeranhatken");
+    question_bounty bountytable(_self, scope_all_bounties);
+    auto iter_bounty = bountytable.find(question_id);
+    eosio::check(iter_bounty->status == BOUNTY_STATUS_ACTIVE,
```
Comment (sergeyi23) [inline src/contracts/peeranha.token/token.cpp:283]: I do not see a check here that the user's answer is accepted as the best for the question.

**Review by sergeyi23 — CHANGES_REQUESTED** [review]
Please see comments.
Also do not see the logic that returns user bounty if delete a question.

**Comment by sergeyi23** [comment]
Also, I hoped that bounty can be set in the same action that posts question.
Let's talk about tables that can be reused between token and main contract.

File: src/contracts/peeranha.main/question_container.hpp (hunk 1 above)
Comment (almondy22) [reply src/contracts/peeranha.main/question_container.hpp:31]: These values were used for integer properties, now we don't need them. I'll remove them.

File: src/contracts/peeranha.token/token.cpp (hunk 3 above)
Comment (almondy22) [reply src/contracts/peeranha.token/token.cpp:275]: Left them for myself and forgot to delete

File: src/contracts/peeranha.token/token.cpp [hunk 7]
```diff
@@ -262,24 +262,40 @@ void token::setbounty(name user, asset bounty, uint64_t question_id, uint64_t ti
...
+        a.amount = bounty;
         a.question_id = question_id;
-        a.timestamp = timestamp;
         a.status = BOUNTY_STATUS_ACTIVE;
+        a.timestamp = timestamp;
     });
 }
 
-void token::paybounty(name user, uint64_t question_id) {
+void token::paybounty(name user, uint64_t question_id, uint64_t best_answer_id, bool on_delete) {
```
Comment (sergeyi23) [inline src/contracts/peeranha.token/token.cpp:275]: why do we need to pass best answer id as an argument? can it be read from the question table?
this opens possibilities for hacks.

File: src/contracts/peeranha.token/token.cpp [hunk 8]
```diff
@@ -262,24 +262,40 @@ void token::setbounty(name user, asset bounty, uint64_t question_id, uint64_t ti
...
+    auto iter_bounty = bounty_table.find(question_id);
+
+    question_index question_table(peeranha_main, scope_all_questions);
+    auto iter_question = question_table.find(question_id);
+    eosio::check(iter_question != question_table.end(), "Question not found!");
+
+    auto iter_answer = binary_find(iter_question->answers.begin(),
+                                   iter_question->answers.end(), best_answer_id);
+
+    if (on_delete == true && iter_question->answers.empty() == true) {
```
Comment (sergeyi23) [inline src/contracts/peeranha.token/token.cpp:287]: what if we forget to call this action and delete a question?

**Review by sergeyi23 — CHANGES_REQUESTED** [review]
see new comments

File: src/contracts/peeranha.token/token.cpp (hunk 8 above)
Comment (almondy22) [reply src/contracts/peeranha.token/token.cpp:287]: What does it mean forget? This action is called in delete question worker in saga. Two actions called together in process of deleting question as it made with post question.

File: src/contracts/peeranha.main/peeranha_forum.cpp [hunk 9]
```diff
@@ -181,9 +182,13 @@ void peeranha::post_comment(eosio::name user, uint64_t question_id,
 void peeranha::delete_question(eosio::name user, uint64_t question_id) {
   auto iter_account = find_account(user);
   auto iter_question = find_question(question_id);
+  question_bounty bounty_table(peeranha_token, scope_all_bounties);
+  auto iter_bounty = bounty_table.find(question_id);
+
   assert_allowed(*iter_account, iter_question->user, Action::DELETE_QUESTION, iter_question->community_id);
   eosio::check(iter_question->answers.empty(),
                "You can't delete not empty question");
+  eosio::check(iter_bounty->status == BOUNTY_STATUS_PAID, "You cannot delete question until bounty is paid");
```
Comment (sergeyi23) [inline src/contracts/peeranha.main/peeranha_forum.cpp:191]: Do we need a check here if bounty does not exist?

File: src/contracts/peeranha.token/token.cpp [hunk 10]
```diff
@@ -256,6 +256,48 @@ void token::inviteuser(name inviter, name invited_user) {
       });
 }
 
+void token::setbounty(name user, asset bounty, uint64_t question_id, uint64_t timestamp) {
+    require_auth(user);
```
Comment (sergeyi23) [inline src/contracts/peeranha.token/token.cpp:260]: Need a check that bounty does not exist yet.

File: src/contracts/peeranha.token/token.cpp [hunk 11]
```diff
@@ -256,6 +256,48 @@ void token::inviteuser(name inviter, name invited_user) {
...
+    auto iter_bounty = bounty_table.find(question_id);
+
+    question_index question_table(peeranha_main, scope_all_questions);
+    auto iter_question = question_table.find(question_id);
+    eosio::check(iter_question != question_table.end(), "Question not found!");
+
+    auto iter_answer = binary_find(iter_question->answers.begin(),
+                                   iter_question->answers.end(), iter_question->correct_answer_id);
+
+    if (on_delete == true && iter_question->answers.empty() == true) {
```
Comment (sergeyi23) [inline src/contracts/peeranha.token/token.cpp:287]: If can be simplified `if (on_delete && iter_question->answers.empty())`

File: src/contracts/peeranha.token/token.cpp [hunk 12]
```diff
@@ -256,6 +256,48 @@ void token::inviteuser(name inviter, name invited_user) {
...
+void token::paybounty(name user, uint64_t question_id, bool on_delete) {
+    require_auth(user);
+    question_bounty bounty_table(_self, scope_all_bounties);
+    auto iter_bounty = bounty_table.find(question_id);
+
+    question_index question_table(peeranha_main, scope_all_questions);
+    auto iter_question = question_table.find(question_id);
+    eosio::check(iter_question != question_table.end(), "Question not found!");
+
+    auto iter_answer = binary_find(iter_question->answers.begin(),
```
Comment (sergeyi23) [inline src/contracts/peeranha.token/token.cpp:284]: Need a check that there is a correct answer before trying to find the correct answer?

File: src/contracts/peeranha.token/token.cpp [hunk 13]
```diff
@@ -256,6 +256,50 @@ void token::inviteuser(name inviter, name invited_user) {
...
+
+    question_index question_table(peeranha_main, scope_all_questions);
+    auto iter_question = question_table.find(question_id);
+    eosio::check(iter_question != question_table.end(), "Question not found!");
+
+    if (on_delete && iter_question->answers.empty()) {
+        eosio::check(iter_question->user == user, "You can't get this bounty");
+        add_balance(user, iter_bounty->amount, user);
+        bounty_table.modify(iter_bounty, _self, [&](auto &a) { a.status = BOUNTY_STATUS_PAID; });
+    } else if (on_delete == false) {
```
Comment (sergeyi23) [inline src/contracts/peeranha.token/token.cpp:290]: `if (!on_delete)`

File: src/contracts/peeranha.token/token.cpp [hunk 14]
```diff
@@ -256,6 +256,50 @@ void token::inviteuser(name inviter, name invited_user) {
...
+        eosio::check(iter_question->correct_answer_id != 0, "Correct answer is not chosen!");
+        auto iter_answer = binary_find(iter_question->answers.begin(),
+                                           iter_question->answers.end(), iter_question->correct_answer_id);
+        eosio::check(iter_answer->user == user, "You can't get this bounty");
+        eosio::check(iter_bounty != bounty_table.end(), "Bounty not found!");
```
Comment (sergeyi23) [inline src/contracts/peeranha.token/token.cpp:295]: Check that bounty is missing does not happen for case when delete question.
Move that check to the top after line `auto iter_bounty = bounty_table.find(question_id);`

File: src/contracts/peeranha.token/token.cpp [hunk 15]
```diff
@@ -256,6 +256,50 @@ void token::inviteuser(name inviter, name invited_user) {
...
+        eosio::check(iter_bounty->status == BOUNTY_STATUS_ACTIVE,
```
Comment (sergeyi23) [inline src/contracts/peeranha.token/token.cpp:296]: Status check can also be done only once for both cases

**Review by sergeyi23 — APPROVED** [review]

**Pull request merged**